File: analyzers/tests/SonarAnalyzer.Test/TestCases/ValueTypeShouldImplementIEquatable.cs
using System;
using System.Collections.Generic;

struct MyStruct // Noncompliant {{Implement 'IEquatable<T>' in value type 'MyStruct'.}}
//     ^^^^^^^^
{
}

struct MyCompliantStruct : IEquatable<MyCompliantStruct> // Compliant
{
    public bool Equals(MyCompliantStruct other)
    {
        throw new NotImplementedException();
    }
}

struct ComparableStruct : IComparable<ComparableStruct> // Noncompliant
{
    public int CompareTo(ComparableStruct other) { return 0; }
}

// https://github.com/SonarSource/sonar-dotnet/issues/3157
ref struct Repro_3157 // Compliant, ref structs can not implement interfaces
{
}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/ValueTypeShouldImplementIEquatable.vb
Public Structure MyCompliantStruct ' Compliant
    Implements IEquatable(Of MyCompliantStruct)

    Public Overloads Function Equals(other As MyCompliantStruct) As Boolean Implements IEquatable(Of MyCompliantStruct).Equals
        Return True
    End Function
End Structure

Structure MyStruct ' Noncompliant {{Implement 'IEquatable<T>' in value type 'MyStruct'.}}
    '     ^^^^^^^^
End Structure

Structure ComparableStruct ' Noncompliant
    Implements IComparable(Of ComparableStruct)

    Public Function CompareTo(other As ComparableStruct) As Integer Implements IComparable(Of ComparableStruct).CompareTo
        Return 0
    End Function
End Structure
